### PR TITLE
Add copy constructor for dynamic_format_arg_store

### PR DESCRIPTION
**Pull request by spyridon97**
The goal of the PR is to develop a copy constructor (and default constructor) for the ``dynamic_format_arg_store``.

The reason behind the need for a copy-constructor as stated in https://github.com/fmtlib/fmt/issues/2431 is that I would like to be able to instantiate a dynamic_format_arg_store object using another one to emulate a scope of arguments based on the class hierarchy of my code.

For the implementation of the copy-constructor, the members: ``data_`` and ``named_info_`` have been copied, while ``dynamic_args_`` remained untouched because it also did not have a copy constructor or an iterator to implement a copy constructor.

Using the implementation included in this PR i was able to successfully do something like below:

```c++
        fmt::dynamic_format_arg_store<fmt::format_context> dArgs;

        std::vector<int> v1 = {5, 1235, 1235123};
        dArgs.push_back(fmt::arg("global_username", std::ref("utkarsh1")));
        dArgs.push_back(fmt::arg("global_hostname", std::ref("miron")));
        dArgs.push_back(fmt::arg("global_vector", v1));

        fmt::dynamic_format_arg_store<fmt::format_context> dArgs2(dArgs);
        dArgs2.push_back(fmt::arg("extra", std::ref("extra-value")));
```

Because of the new copy constructor, ``dArgs`` includes the first 3 variables, and ``dArgs2`` includes the first 3 variables plus the 4th one.

### CODE REVIEW

File: include/fmt/args.h [hunk 1]
```diff
@@ -143,6 +143,38 @@ class dynamic_format_arg_store
   }
 
  public:
+  /**
+    \rst
+    Default constructor of `dynamic_format_arg_store`
+    \endrst
+  */
```
Comment (vitaut) [inline include/fmt/args.h:150]: The comment doesn't say anything useful, please remove.

File: include/fmt/args.h [hunk 2]
```diff
@@ -143,6 +143,38 @@ class dynamic_format_arg_store
...
+  constexpr FMT_INLINE dynamic_format_arg_store() = default;
```
Comment (vitaut) [inline include/fmt/args.h:151]: `FMT_INLINE` is not needed here.

File: include/fmt/args.h [hunk 3]
```diff
@@ -143,6 +143,38 @@ class dynamic_format_arg_store
...
+
+    **Example**::
+
+      fmt::dynamic_format_arg_store<fmt::format_context> store;
+      store.push_back(123);
+      store.push_back("test");
+
+      fmt::dynamic_format_arg_store<fmt::format_context> store2(store);
+    \endrst
+  */
```
Comment (vitaut) [inline include/fmt/args.h:166]: Let's remove this too because we don't need to explain what a copy constructor is.

File: include/fmt/args.h [hunk 4]
```diff
@@ -143,6 +143,38 @@ class dynamic_format_arg_store
...
+  constexpr FMT_INLINE dynamic_format_arg_store(
```
Comment (vitaut) [inline include/fmt/args.h:167]: `FMT_INLINE` is not needed here.

File: include/fmt/args.h [hunk 5]
```diff
@@ -143,6 +143,38 @@ class dynamic_format_arg_store
...
+      const dynamic_format_arg_store<Context>& store)
+      : data_(store.data_),
+        named_info_(store.named_info_)
+#if FMT_GCC_VERSION && FMT_GCC_VERSION < 409
+        ,
+        basic_format_args<Context>()
+#endif
```
Comment (vitaut) [inline include/fmt/args.h:174]: Why is this necessary?

**Review by vitaut — CHANGES_REQUESTED** [review]
Thanks for the PR. A few small comments inline.

**Comment by vitaut** [comment]
Also please add a test case to https://github.com/fmtlib/fmt/blob/master/test/args-test.cc.

File: include/fmt/args.h (hunk 5 above)
Comment (spyridon97) [reply include/fmt/args.h:174]: this was necessary(?) because it way saying that the copy constructor needs to initialize the parent class too.

Comment (vitaut) [reply include/fmt/args.h:174]: I suggest moving this to the beginning of the initialization list to improve formatting and since base classes are normally initialized first.

File: include/fmt/args.h [hunk 6]
```diff
@@ -230,3 +242,4 @@ class dynamic_format_arg_store
 FMT_END_NAMESPACE
 
 #endif  // FMT_ARGS_H_
+
```
Comment (vitaut) [inline include/fmt/args.h:245]: Please remove the extra newline.

File: test/args-test.cc [hunk 7]
```diff
@@ -171,3 +171,16 @@ TEST(args_test, throw_on_copy) {
...
   EXPECT_EQ(fmt::vformat("{}", store), "foo");
 }
+
+TEST(args_test, copy_constructor) {
+  auto store = fmt::dynamic_format_arg_store<fmt::format_context>();
+  store.push_back(fmt::arg("test1", "value1"));
+  store.push_back(fmt::arg("test2", "value2"));
+  store.push_back(fmt::arg("test3", "value3"));
+
+  fmt::dynamic_format_arg_store<fmt::format_context> store2(store);
```
Comment (vitaut) [inline test/args-test.cc:181]: nit: let's use the same declaration format here

```c++
auto store2 = fmt::dynamic_format_arg_store<fmt::format_context>(store);
```

**Review by vitaut — CHANGES_REQUESTED** [review]

File: include/fmt/args.h (hunk 6 above)
Comment (spyridon97) [reply include/fmt/args.h]: Fixed

File: test/args-test.cc (hunk 7 above)
Comment (spyridon97) [reply test/args-test.cc]: also fixed

File: test/args-test.cc [hunk 8]
```diff
@@ -171,3 +171,16 @@ TEST(args_test, throw_on_copy) {
...
   EXPECT_EQ(fmt::vformat("{}", store), "foo");
 }
+
+TEST(args_test, copy_constructor) {
+  auto store = fmt::dynamic_format_arg_store<fmt::format_context>();
+  store.push_back(fmt::arg("test1", "value1"));
+  store.push_back(fmt::arg("test2", "value2"));
+  store.push_back(fmt::arg("test3", "value3"));
+
+  auto store2 = fmt::dynamic_format_arg_store<fmt::format_context>(store);
```
Comment (vitaut) [inline test/args-test.cc:181]: To get rid of the "useless cast" gcc warning:

```c++
  auto store2 = store;
```

Comment (spyridon97) [reply test/args-test.cc]: Fixed too.

**Comment by spyridon97** [comment]
AppVeyor build succeeded

**Review by vitaut — APPROVED** [review]

**Comment by vitaut** [comment]
@DanielaE, do you know why msvc complains about `test-module.ifc`?

**Comment by DanielaE** [comment]
The good news: Spiros' changes are fine, everything builds and all tests pass with msvc 16.10, 16.11-pre3 and 17.0-pre2. 🎉

The bad news: CI has a new windows-2019 image with CMake 3.21 instead of 3.20. The compiler is still the same (16.10.2)
The error message looks like the command line parameters for compiling `fmt.cc` as a module interface unit are missing with CMake 3.21. That's just a guess derived from the error messages without knowing the actual command line parameters in the compiler invocation! The builds on my local machine are still on CMake 3.20 (plus Ninja as supplied by msvc) where everything is fine with the default 26 parallel build threads in VSCode.

I will probably not find enough time today though to switch my local build environment to CMake 3.21 and look into differences in `build.ninja` generated by 3.20 and 3.21 🤔 Stay tuned!

**Comment by DanielaE** [comment]
Everything builds locally with CMake 3.21! 😱
The compiler options are the same as before. But everything is switched from relative paths to absolute paths in `build.ninja`.
I have no clue how to proceed from here without seeing details and related files from the CI build ...

**Comment by spyridon97** [comment]
> Everything builds locally with CMake 3.21! 
> The compiler options are the same as before. But everything is switched from relative paths to absolute paths in `build.ninja`.
> I have no clue how to proceed from here without seeing details and related files from the CI build ...

**Comment by spyridon97** [comment]
I accidentally closed this PR, so I reopened it. 

Is there any progress as far as the Cmake issues? @DanielaE

**Comment by DanielaE** [comment]
I just found out that the CMake generators "Visual Studio 16 2019" and "Visual Studio 17 2022" are 'improved' in CMake 3.21 in the sense that they are broken: they swallow the compiler option `/interface` in `CMakeLists.txt` line 31
```
  target_compile_options(${target}
      PRIVATE /interface /ifcOutput ${BMI}
      INTERFACE /reference fmt=${BMI})
```
Whatever I've tried to hide these compiler options from being processed by the generators fell flat. `/interface` is removed and `/ifcOutput ...` is converted into `<ModuleOutputFile> ./fmt/build/test/test-module.ifc</ModuleOutputFile>` in `build/test/test-module.vcxproj`. The latter is ok, but the former is 😱

**Comment by vitaut** [comment]
In that case I suggest temporarily setting `FMT_CAN_MODULE` to `OFF` in https://github.com/fmtlib/fmt/blob/561834650aa77ba37b15f7e5c9d5726be5127df9/CMakeLists.txt#L89 to unblock this PR.

**Comment by vitaut** [comment]
Thank you!

**Comment by DanielaE** [comment]
> I just found out that the CMake generators "Visual Studio 16 2019" and "Visual Studio 17 2022" are 'improved' in CMake 3.21 in the sense that they are broken: they swallow the compiler option `/interface` in `CMakeLists.txt` line 31

A fix for that will hopefully land in CMake 3.21.1:
- [Issue](https://gitlab.kitware.com/cmake/cmake/-/issues/22477)
- [Direction](https://gitlab.kitware.com/cmake/cmake/-/issues/22474)

**Comment by vitaut** [comment]
Thanks, @DanielaE, for looking into it!

**Comment by spyridon97** [comment]
@vitaut I saw in the newest version of fmt this copy-constructor was deleted. Would it be possible to explain why?

**Comment by vitaut** [comment]
IIRC it was broken, you can look up the commit history for more details.

**Comment by spyridon97** [comment]
I looked it up, there was not MR for the commit that did that (https://github.com/fmtlib/fmt/commit/be51ee1ceb0553411cae166335b59e7ae5126527)

**Comment by cdacamar** [comment]
@spyridon97 do you know if the CMake issue was addressed which ran afoul with IFC generation?

**Comment by spyridon97** [comment]
> @spyridon97 do you know if the CMake issue was addressed which ran afoul with IFC generation?

Issue: https://gitlab.kitware.com/cmake/cmake/-/issues/22477 has been merged
Issue: https://gitlab.kitware.com/cmake/cmake/-/issues/22474 has a discussion which you can check
